grpcbp/server.go

```diff
@@ -0,0 +1,128 @@
+	// connection storms. The current default value is infinity.
+	MaxConnectionAge time.Duration `yaml:"maxConnectionAge"`
+
+	// MaxConnectionAgeGrace is an additive period after MaxConnectionAge after
+	// which the connection will be forcibly closed. The current defualt value is
+	// infinity.
+	MaxConnectionAgeGrace time.Duration `yaml:"maxConnectionAgeGrace"`
+
+	// After a duration of this time if the server doesn't see any activity it
+	// pings the client to see if the transport is still alive. If set below 1s,
+	// a minimum value of 1s will be used instead. The current default value is 2
+	// hours.
+	Time time.Duration `yaml:"time"`
+
+	// After having pinged for keepalive check, the server waits for a duration
+	// of Timeout and if no activity is seen even after that the connection is
+	// closed. The current default value is 20 seconds.
+	Timeout time.Duration `yaml:"timeout"`
+
+	// MinTime is the minimum amount of time a client should wait before sending
+	// a keepalive ping.
+	MinTime time.Duration `yaml:"minTime"`
+
+	// If true, server allows keepalive pings even when there are no active
+	// streams(RPCs). If false, and client sends ping when there are no active
+	// streams, server will send GOAWAY and close the connection.
+	PermitWithoutStream bool `yaml:"permitWithoutStream"`
+
+	// RegisterServerFunc is used to pass in a generated gRPC service
+	// implementation and register it on the created gRPC.Server.
+	RegisterServerFunc func(*grpc.Server) `yaml:"-"`
+}
+
+// NewBaseplateServer returns a new gRPC implementation of a Baseplate server
+// with the given config.
+func NewBaseplateServer(bp baseplate.Baseplate, cfg ServerConfig) (baseplate.Server, error) {
+	lis, err := net.Listen("tcp", bp.GetConfig().Addr)
+	if err != nil {
+		return nil, err
+	}
+
+	kaep := keepalive.EnforcementPolicy{
+		MinTime:             cfg.MinTime,
+		PermitWithoutStream: cfg.PermitWithoutStream,
+	}
+
+	kasp := keepalive.ServerParameters{
+		MaxConnectionIdle:     cfg.MaxConnectionIdle,
+		MaxConnectionAge:      cfg.MaxConnectionAge,
+		MaxConnectionAgeGrace: cfg.MaxConnectionAgeGrace,
+		Time:                  cfg.Time,
+		Timeout:               cfg.Timeout,
+	}
+
+	middlewares := BaseplateDefaultMiddlewares(DefaultMiddlewaresArgs{
+		EdgeContextImpl: bp.EdgeContextImpl(),
+	})
+
+	server := grpc.NewServer(
+		grpc.KeepaliveEnforcementPolicy(kaep),
+		grpc.KeepaliveParams(kasp),
+		middlewares,
+	)
+	cfg.RegisterServerFunc(server)
+
+	return ApplyBaseplate(bp, server, lis), nil
+}
+
+// ApplyBaseplate returns the given grpc.Server as a baseplate server with the
+// provided Baseplate.
+//
+// You generally don't need to use this, instead use NewBaseplateServer, which
+// will take care of this for you.
+func ApplyBaseplate(bp baseplate.Baseplate, server *grpc.Server, lis net.Listener) baseplate.Server {
+	return impl{
+		bp:  bp,
+		lis: lis,
+		srv: server,
+	}
+}
+
+type impl struct {
```

Comment:
"impl" is a no-name name.  "server" or "grpcServer" would be better names for this, since it's implementing baseplate.Server